### PR TITLE
AWS-MSK implementation updates

### PR DESCRIPTION
**Pull request by bpathak-ons**
### What

* Enable app to use Kafka TLS / AWS-MSK
* Add the relevant AWS-MSK notes, configs and code

### How to review
* check if code changes make sense
* CI pass for build, audit and unit

### Who can review
* anyone

### CODE REVIEW

File: service/initialise.go [hunk 1]
```diff
@@ -65,15 +68,27 @@ func (e *Init) DoGetHTTPServer(bindAddr string, router http.Handler) HTTPServer
...
 	}
 	kafkaConsumer, err := dpkafka.NewConsumerGroup(
 		ctx,
 		cfg.KafkaAddr,
-		cfg.HelloCalledTopic,
-		cfg.HelloCalledGroup,
+		cfg.PublishedContentTopic,
+		cfg.PublishedContentGroup,
 		cgChannels,
 		&dpkafka.ConsumerGroupConfig{
```
Comment (gedge) [inline service/initialise.go:93]: this should be using `cConfig` from above?

File: service/initialise.go [hunk 2]
```diff
@@ -65,15 +68,27 @@ func (e *Init) DoGetHTTPServer(bindAddr string, router http.Handler) HTTPServer
 // DoGetKafkaConsumer returns a Kafka Consumer group
 func (e *Init) DoGetKafkaConsumer(ctx context.Context, cfg *config.Config) (dpkafka.IConsumerGroup, error) {
 	cgChannels := dpkafka.CreateConsumerGroupChannels(1)
+
+	cConfig := &dpkafka.ConsumerGroupConfig{
+		KafkaVersion: &cfg.KafkaVersion,
```
Comment (gedge) [inline service/initialise.go:73]: this should include `kafkaOffset` from below

File: cmd/producer/main.go [hunk 3]
```diff
@@ -29,11 +29,11 @@ func main() {
 
 	// Create Kafka Producer
 	pChannels := kafka.CreateProducerChannels()
-	kafkaProducer, err := kafka.NewProducer(ctx, config.KafkaAddr, config.HelloCalledTopic, pChannels, &kafka.ProducerConfig{
+	kafkaProducer, err := kafka.NewProducer(ctx, config.KafkaAddr, config.PublishedContentTopic, pChannels, &kafka.ProducerConfig{
```
Comment (gedge) [inline cmd/producer/main.go:32]: this should be using something like `pConfig` which would be created beforehand and optionally includes the `SecurityConfig` for TLS

File: config/config.go [hunk 4]
```diff
@@ -6,19 +6,23 @@ import (
...
 
-// TODO: remove hello call config options
 // Config represents service configuration for dp-search-data-importer
 type Config struct {
 	BindAddr                   string        `envconfig:"BIND_ADDR"`
 	GracefulShutdownTimeout    time.Duration `envconfig:"GRACEFUL_SHUTDOWN_TIMEOUT"`
 	HealthCheckInterval        time.Duration `envconfig:"HEALTHCHECK_INTERVAL"`
 	HealthCheckCriticalTimeout time.Duration `envconfig:"HEALTHCHECK_CRITICAL_TIMEOUT"`
-	KafkaAddr                  []string      `envconfig:"KAFKA_ADDR"                     json:"-"`
+	KafkaAddr                  []string      `envconfig:"KAFKA_ADDR"                    json:"-"`
```
Comment (gedge) [inline config/config.go:15]:
```suggestion
	KafkaAddr                  []string      `envconfig:"KAFKA_ADDR"`
```
no need to hide this

File: README.md [hunk 5]
```diff
@@ -17,17 +17,26 @@ An example event can be created using the helper script, `make produce`.
...
-| KAFKA_NUM_WORKERS            | 1                                 | The maximum number of parallel kafka consumers
-| HELLO_CALLED_GROUP           | dp-search-data-importer           | The consumer group this application to consume ImageUploaded messages
-| HELLO_CALLED_TOPIC           | hello-called                      | The name of the topic to consume messages from
+| Environment variable          | Default                           | Description
+| ----------------------------  | --------------------------------- | -----------
+| BIND_ADDR                     | localhost:25900                   | The host and port to bind to
+| GRACEFUL_SHUTDOWN_TIMEOUT     | 5s                                | The graceful shutdown timeout in seconds (`time.Duration` format)
+| HEALTHCHECK_INTERVAL          | 30s                               | Time between self-healthchecks (`time.Duration` format)
+| HEALTHCHECK_CRITICAL_TIMEOUT  | 90s                               | Time to wait until an unhealthy dependent propagates its state to make this app unhealthy (`time.Duration` format)
+| KAFKA_ADDR                    | "localhost:9092"                  | The address of Kafka (accepts list)
```
Comment (gedge) [inline README.md:26]: this should have `KAFKA_VERSION` too

File: cmd/producer/main.go [hunk 6]
```diff
@@ -16,6 +16,9 @@ import (
 
 const serviceName = "dp-search-data-importer"
 
+// KafkaTLSProtocolFlag informs service to use TLS protocol for kafka
+const KafkaTLSProtocolFlag = "TLS"
```
Comment (gedge) [inline cmd/producer/main.go:20]: I would argue that this should go into `config` package

File: cmd/producer/main.go [hunk 7]
```diff
@@ -29,11 +32,21 @@ func main() {
 
 	// Create Kafka Producer
 	pChannels := kafka.CreateProducerChannels()
-	kafkaProducer, err := kafka.NewProducer(ctx, config.KafkaAddr, config.HelloCalledTopic, pChannels, &kafka.ProducerConfig{
+	pConfig := &kafka.ProducerConfig{
 		KafkaVersion: &config.KafkaVersion,
-	})
+	}
+	if config.KafkaSecProtocol == KafkaTLSProtocolFlag {
```
Comment (gedge) [inline cmd/producer/main.go:38]:
```suggestion
	if config.KafkaSecProtocol == config.KafkaTLSProtocolFlag {
```

File: service/initialise.go [hunk 8]
```diff
@@ -10,6 +10,9 @@ import (
 	"github.com/ONSdigital/dp-search-data-importer/config"
 )
 
+// KafkaTLSProtocolFlag informs service to use TLS protocol for kafka
+const KafkaTLSProtocolFlag = "TLS"
```
Comment (gedge) [inline service/initialise.go:14]: remove this (assuming you put it in `config`)

File: service/initialise.go [hunk 9]
```diff
@@ -65,15 +68,30 @@ func (e *Init) DoGetHTTPServer(bindAddr string, router http.Handler) HTTPServer
...
 	kafkaOffset := dpkafka.OffsetNewest
 	if cfg.KafkaOffsetOldest {
 		kafkaOffset = dpkafka.OffsetOldest
 	}
+
+	cConfig := &dpkafka.ConsumerGroupConfig{
+		KafkaVersion: &cfg.KafkaVersion,
+		Offset: &kafkaOffset,
+	}
+	if cfg.KafkaSecProtocol == "TLS" {
```
Comment (gedge) [inline service/initialise.go:81]:
```suggestion
	if cfg.KafkaSecProtocol == config.KafkaTLSProtocolFlag {
```

File: service/initialise.go (hunk 1 above)
Comment (gedge) [reply service/initialise.go:96]: this remains to be done

**Review by justinpjose — DISMISSED** [review]

File: config/config.go [hunk 10]
```diff
@@ -6,19 +6,26 @@ import (
 	"github.com/kelseyhightower/envconfig"
 )
 
-// TODO: remove hello call config options
+// KafkaTLSProtocolFlag informs service to use TLS protocol for kafka
+const KafkaTLSProtocolFlag = "TLS"
```
Comment (nshumoogum) [inline config/config.go:10]: Im not sure why this is needed?

Default is empty string `""` and then if we want TLS then we set the `KAFKA_SEC_PROTO` to be `"TLS"`

Comment (nshumoogum) [reply config/config.go:10]: Okay further looking into this - this isn't a flag, it just describes a kafka protocol that is used to compare against the actual flag `KafkaSecProtocol`.

The description doesn't help decipher what this actually is.

```suggestion
// KafkaTLSProtocol is a constant describing the TLS protocol used for kafka
const KafkaTLSProtocol = "TLS"
```

**Review by nshumoogum — DISMISSED** [review]

**Review by justinpjose — APPROVED** [review]

**Pull request merged**